crates/sui-framework/packages/move-stdlib/sources/vector.move

```diff
@@ -158,4 +158,108 @@ module std::vector {
         v.swap(i, last_idx);
         v.pop_back()
     }
+
+    // === Macros ===
+
+    /// Destroy the vector `v` by calling `f` on each element and then destroying the vector.
+    public macro fun destroy<$T>($v: vector<$T>, $f: |$T|) {
+        let mut v = $v;
+        while (!v.is_empty()) {
+            $f(v.pop_back());
+        };
+        v.destroy_empty();
+    }
+
+    /// Map the vector `v` to a new vector by applying the function `f` to each element.
+    public macro fun map<$T, $U>($v: vector<$T>, $f: |$T| -> $U): vector<$U> {
+        let mut r = vector[];
+        let mut v = $v;
+        while (!v.is_empty()) {
+            r.push_back($f(v.pop_back()));
+        };
+        v.destroy_empty();
+        r.reverse();
+        r
+    }
+
+    /// Map the vector `v` to a new vector by applying the function `f` to each element.
+    public macro fun map_ref<$T, $U>($v: &vector<$T>, $f: |&$T| -> $U): vector<$U> {
+        let v = $v;
+        let mut r = vector[];
+        let mut i = 0;
+        let len = v.length();
+        while (i < len) {
+            r.push_back($f(&v[i]));
+            i = i + 1;
+        };
+        r
+    }
+
+    /// Filter the vector `v` by applying the function `f` to each element.
+    /// Return a new vector containing only the elements for which `f` returns `true`.
+    public macro fun filter<$T: drop>($v: vector<$T>, $f: |&$T| -> bool): vector<$T> {
+        let mut r = vector[];
+        let mut v = $v;
+        while (!v.is_empty()) {
+            let e = v.pop_back();
+            if ($f(&e)) r.push_back(e);
+        };
+        v.destroy_empty();
+        r.reverse();
+        r
+    }
+
+    /// Split the vector `v` into two vectors by applying the function `f` to each element.
+    /// Return a tuple containing two vectors: the first containing the elements for which `f` returns `true`,
+    /// and the second containing the elements for which `f` returns `false`.
+    public macro fun partition<$T>($v: vector<$T>, $f: |&$T| -> bool): (vector<$T>, vector<$T>) {
+        let mut r1 = vector[];
+        let mut r2 = vector[];
+        let mut v = $v;
+        while (!v.is_empty()) {
+            let e = v.pop_back();
+            if ($f(&e)) r1.push_back(e)
+            else r2.push_back(e);
+        };
+        v.destroy_empty();
+        r1.reverse();
+        r2.reverse();
+        (r1, r2)
+    }
+
+    /// Reduce the vector `v` to a single value by applying the function `f` to each element.
+    /// Similar to `fold_left` in Rust and `reduce` in Python and JavaScript.
+    public macro fun fold<$T, $Acc>($v: vector<$T>, $init: $Acc, $f: |$Acc, $T| -> $Acc): $Acc {
```

Comment:
Consider swapping the order of arguments for `init` to indicate the fold direction (the accumulator is on the "right" in the folded tree).

Reply:
Are you suggesting to take `$Acc` as the last argument?

Reply:
Also, do you think we should reverse in this case too?